[reland] Catch saved tensors default hooks race condition

Varal7 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62564

If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

See [Benchmark](https://github.com/pytorch/pytorch/pull/62564#issuecomment-891242044)

Relanding previous PR #61957

Differential Revision: [D30045406](https://our.internmc.facebook.com/intern/diff/D30045406)

Varal7 committed:
[reland] Catch saved tensors default hooks race condition

If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -45,12 +45,12 @@ namespace torch { namespace autograd {
+  PyObject* PyDefaultSavedVariableHooks::unpack_hook_(nullptr);
 
   void PyDefaultSavedVariableHooks::set_hooks(py::function &pack_hook, py::function &unpack_hook) {
+    std::lock_guard<std::mutex> lock(mutex_);
```
For this to be effective, you need to protect every function that changes these hooks so the reset below in particular.

Also if you don't protect the get_hook, you can still get in a situation where a user can get a pack hook from one pair and an unpack from another (or one empty). Which is bad as well.

Varal7 committed:
Update on "[reland] Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_saved_variable_hooks.cpp:
```diff
@@ -69,6 +70,7 @@ namespace torch { namespace autograd {
   }
 
   std::unique_ptr<SavedVariableHooks> PyDefaultSavedVariableHooks::get_hooks() {
+    std::lock_guard<std::mutex> lock(mutex_);
```
nit:
I think we can move this one after the if statement.
Keeping in mind the heavily used case where every SavedVariable creation will call here to check if there are default hooks. If we can avoid taking the lock every time it is better.

And here we are happy with returning nothing even if another thread is filling/deleting things.
Only the part where we actually read their content needs to be protected.

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30045406).

Varal7 commented:
Benchmark!

**On master** (before #62563)
Forward | Backward
-- | --
0.4018306732 | 0.850032568
0.3812365532 | 0.8270413876
0.4073872566 | 0.8165071011
0.3809463978 | 0.8159003258
0.3936896324 | 0.9021894932
0.4049458504 | 0.8370418549
0.3917701244 | 0.8697845936
0.4148375988 | 0.8651194572
0.3916740417 | 0.8183619976
0.3931877613 | 0.8848567009

**After this patch**
0.396 +/- 0.011 | 0.849 +/- 0.031
 
Forward | Backward
-- | --
0.3958051205 | 0.8655035496
0.4148197174 | 0.9240355492
0.4204208851 | 0.8634562492
0.3910806179 | 0.8600254059
0.3905735016 | 0.8833994865
0.4093549252 | 0.909497261
0.4033236504 | 0.848752737
0.3930876255 | 0.8423700333
0.4047224522 | 0.8782734871
0.4443187714 | 0.8477358818

0.407 +/- 0.017 | 0.872 +/- 0.027

Varal7 committed:
Update on "[reland] Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

See [Benchmark](https://github.com/pytorch/pytorch/pull/62564#issuecomment-891242044)

Relanding previous PR #61957

Differential Revision: [D30045406](https://our.internmc.facebook.com/intern/diff/D30045406)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30045406).

albanD reviewed (approved):
LGTM

Varal7 committed:
Update on "[reland] Catch saved tensors default hooks race condition"


If the user runs code that registers default saved tensor hooks from
multiple threads, it will fail with a nice error message most of the
time. This commit handles the very rare case where a race condition
would have made it fail silently.

See [Benchmark](https://github.com/pytorch/pytorch/pull/62564#issuecomment-891242044)

Relanding previous PR #61957

Differential Revision: [D30045406](https://our.internmc.facebook.com/intern/diff/D30045406)

[ghstack-poisoned]

Varal7 commented:
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30045406).
